Code/Source/Spawner/ROS2SpawnerComponent.cpp

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include "ROS2SpawnerComponent.h"
+#include <AzCore/Math/Quaternion.h>
+#include <AzCore/Serialization/EditContext.h>
+#include <AzFramework/Components/TransformComponent.h>
+#include <AzToolsFramework/UI/PropertyEditor/PropertyEditorAPI.h>
+
+namespace ROS2
+{
+    void ROS2SpawnerComponent::Activate()
+    {
+        auto ros2Node = ROS2Interface::Get()->GetNode();
+
+        m_get_names_service = ros2Node->create_service<gazebo_msgs::srv::GetWorldProperties>(
+            "get_available_spawnable_names",
+            std::bind(&ROS2SpawnerComponent::GetAvailableSpawnableNames, this, AZStd::placeholders::_1, AZStd::placeholders::_2));
+
+        m_spawn_service = ros2Node->create_service<gazebo_msgs::srv::SpawnEntity>(
+            "spawn_entity", std::bind(&ROS2SpawnerComponent::SpawnRobot, this, AZStd::placeholders::_1, AZStd::placeholders::_2));
+    }
+
+    void ROS2SpawnerComponent::Deactivate()
+    {
+        m_get_names_service.reset();
+        m_spawn_service.reset();
+    }
+
+    void ROS2SpawnerComponent::Reflect(AZ::ReflectContext* context)
+    {
+        if (AZ::SerializeContext* serialize = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serialize->Class<ROS2SpawnerComponent, AZ::Component>()->Version(1)->Field("Spawnables", &ROS2SpawnerComponent::m_spawnables);
+
+            if (AZ::EditContext* ec = serialize->GetEditContext())
+            {
+                ec->Class<ROS2SpawnerComponent>("ROS2 Spawner", "Spawner component")
+                    ->ClassElement(AZ::Edit::ClassElements::EditorData, "Manages spawning of robots in configurable locations")
+                    ->Attribute(AZ::Edit::Attributes::AppearsInAddComponentMenu, AZ_CRC("Game"))
+                    ->DataElement(AZ::Edit::UIHandlers::EntityId, &ROS2SpawnerComponent::m_spawnables, "Spawnables", "Spawnables");
+            }
+        }
+    }
+
+    void ROS2SpawnerComponent::GetAvailableSpawnableNames(
+        const GetAvailableSpawnableNamesRequest request, GetAvailableSpawnableNamesResponse response)
```

Comment:
Both of these parameters should be by reference.
The first parameter is const reference, the second with non-const reference.
```suggestion
    void ROS2SpawnerComponent::GetAvailableSpawnableNames(
        const GetAvailableSpawnableNamesRequest& request, GetAvailableSpawnableNamesResponse& response)
```

Reply:
ROS2 Node::create_service method does not accept the callback with reference arguments. It could work if I switched from bind:
`ros2Node->create_service<gazebo_msgs::srv::GetWorldProperties>("get_available_spawnable_names", std::bind(&ROS2SpawnerComponent::GetAvailableSpawnableNames, this, AZStd::placeholders::_1, AZStd::placeholders::_2));` 
to lamda:
`ros2Node->create_service<gazebo_msgs::srv::GetWorldProperties>("get_available_spawnable_names", [this](const GetAvailableSpawnableNamesRequest req, GetAvailableSpawnableNamesResponse res){ this->GetAvailableSpawnableNames(req, res); })`

In the second case the GetAvailableSpawnableNames signature can look the way you suggested. But I think we would not gain anything from it since the copy is still needed to call the lambda callback? Correct me please if I am wrong.

Reply:
It is also done without the references in [ROS2 service examples](https://docs.ros.org/en/galactic/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Cpp-Service-And-Client.html#write-the-service-node), so I assumed it is the correct way.

Reply:
It looks like [GetAvailableSpawnableNamesResponse](https://github.com/RobotecAI/o3de-ros2-gem/pull/172/files#diff-2f072dcae025d6df3fbcdf9ab9981856ef73e0c68f4b84e770eaf4098d8ec7e7R21) is type alias to a shared_ptr.

I thought this was an raw value object, so it' wasn't being updated.
It makes sense now that ht parameter is passed by value.